propagator/analysis.py

```diff
@@ -362,21 +356,77 @@ def _reduce(_ml, _out_ml):
     _out_ml : str
         File path to an empty temporary feature class that will be
         filled in this function.
+    wq_fields : list
+        Lisf of water quality parameters to be analyzed
+    subcatch_id_col : str
+        Field name of the subcatchment ID column in ``subcatchments``.
 
     Returns
     -------
     _out_ml : str
-
+    
     """
 
-    #load water quality data - place holder for now
-    _arr = utils.load_attribute_table(_ml, 'FID_ml', 'Catch_ID_a')
-
-    # for testing purpose at the moment, export the feature with the
-    # smallest FID as output
-    _sqlexp = '"FID_ml" = ' + "%i" % numpy.min(_arr['FID_ml'])
+    # Load water quality data 
+    _arr = utils.load_attribute_table(_ml, sort_id, *[f for f in wq_fields])
+    
+    # Copy and paste the monitoring location with the smallest FID.
+    # This is essentially an arbiutary pick. What matters is we need
+    # to only output one point back to the upstream function.
+    _sqlexp = '"{}" = {}'.format(sort_id, numpy.min(_arr[sort_id]))
     arcpy.analysis.Select(_ml, _out_ml, _sqlexp)
-
-    #process the WQ data - [e.g. out_ml.wq = average(temp_ml.wq)]
+    #_sqlexp = "%s" + '= ' + "%i" % sort_id, numpy.min(_arr[sort_id])
+    
+    
+    # Loop through each water quality parameter.
+    for wq_par in wq_fields:
+        wq_value  = utils.groupby_and_aggregate(
+            input_path= _ml,
+            groupfield=subcatch_id_col,
+            valuefield=wq_par,
+            # If at futre stage it is decided to compute the water quality 
+            # value with a different methods (i.e. median), either revise 
+            # the existing _non_zero_means function, or (preferrably) write
+            # a new function (i.e. non_zero_median) and call the new function
+            # at the line below.
+            aggfxn= _non_zero_means
+        )
+        
+        # Overwrite field [wq_par] of _out_ml with the computed value.
+        utils.populate_field(_out_ml, lambda row: wq_value.values()[0], wq_par)
     return _out_ml
 
+def _non_zero_means(_arr):
+    """
+    Compute average value of all non-zero values in the 
+    list.
+    
+    Parameters
+    ----------
+    _arr : object 
+        Contains the list of interested values, which 
+        are extracted from the parental feature.
+
+    Returns
+    -------
+    numpy.mean(numlst): float
+        Zero-excluded average of the list. If the list 
+        contains no value, return a value of 0. 
+    """
+    
+    
+    try:
+        # For some reason when this function is called from _reduce, _arr
+        # is not treated as a regular list, and only the following line
+        # will work to correctly extract numeric values from _arr
+        _numlst = filter(lambda n: n>0, [r[1] for r in _arr])
+    except:
+        # This line works when _arr is a regular list (for which case the
```

Comment:
Upon thinking about this, the statement in the `try` block is intended for arcpy Cursors. So really we should just be checking the types, there. e.g., 

```python
if isinstance(_arr, arcpy.Cursor)`:
    # yada yada
elif isinstance(_arr, numpy.ndarray):
    # yada
else:
    raise ValueError("only numpy arrays and arcpy Cursors are supported")
```